runner/internal/shim/docker.go

```diff
@@ -169,38 +178,38 @@ func (d *DockerRunner) Run(ctx context.Context, cfg TaskConfig) error {
 		}
 	}()
 
-	d.state = Running
-	d.containerID = containerID
-	d.containerName = cfg.ContainerName
-	log.Printf("Running container, name=%s, id=%s\n", d.containerName, containerID)
+	log.Printf("Running container, name=%s, id=%s\n", task.containerName, containerID)
+	task.SetStatusRunning(containerID)
+	if !d.tasks.Update(task) {
+		return tracerr.Errorf("failed to update task %s", task.ID)
+	}
 
 	if err = runContainer(ctx, d.client, containerID); err != nil {
 		log.Printf("runContainer error: %s\n", err.Error())
-		d.state = Pending
-		d.containerID = ""
 		var errMessage string
 		if lastLogs, err := getContainerLastLogs(d.client, containerID, 5); err == nil {
 			errMessage = strings.Join(lastLogs, "\n")
 		} else {
 			log.Printf("getContainerLastLogs error: %s\n", err.Error())
 			errMessage = ""
 		}
-		d.jobResult = JobResult{Reason: "CONTAINER_EXITED_WITH_ERROR", ReasonMessage: errMessage}
+		task.SetStatusTerminated("CONTAINER_EXITED_WITH_ERROR", errMessage)
 		return tracerr.Wrap(err)
 	}
 
-	log.Printf("Container finished successfully, name=%s, id=%s", d.containerName, containerID)
-	d.state = Pending
-	d.containerID = ""
-
-	d.jobResult = JobResult{Reason: "DONE_BY_RUNNER"}
+	log.Printf("Container finished successfully, name=%s, id=%s", task.containerName, containerID)
+	task.SetStatusTerminated("DONE_BY_RUNNER", "")
 
 	return nil
 }
 
 func (d *DockerRunner) Stop(force bool) {
-	if d.state == Pulling && d.containerID == "" {
-		d.cancelPull()
+	task, ok := d.tasks.Get(LegacyTaskID)
+	if !ok {
+		return
+	}
+	if task.Status == TaskStatusPulling {
+		task.cancelPull()
 		return
 	}
```

Comment:
Is any handling needed for the pending status too? Otherwise shim will attempt to stop a container that does not exist yet (line 222)